Upstream visionOS fullscreen transition logic

pxlcoder opened the pull request:
#### 98669482d5b2a3d1b747881d8b15bbe32f35bb7b
<pre>
Upstream visionOS fullscreen transition logic
<a href="https://bugs.webkit.org/show_bug.cgi?id=258404">https://bugs.webkit.org/show_bug.cgi?id=258404</a>
rdar://111164029

Reviewed by Wenson Hsieh.

* Source/WebKit/Platform/spi/visionos/MRUIKitSPI.h: Added.
* Source/WebKit/Platform/spi/visionos/RealitySimulationServicesSPI.h: Added.
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(WebKit::useSpatialFullScreenTransition):
(WebKit::resizeScene):
(-[WKFixedSizeWindow _shouldResizeWithScene]):
(-[WKFullScreenParentWindowState initWithWindow:]):
(-[WKFullscreenWindowSceneDelegate initWithController:originalDelegate:]):
(-[WKFullscreenWindowSceneDelegate originalDelegate]):
(-[WKFullscreenWindowSceneDelegate windowScene:shouldCloseForReason:]):
(-[WKFullscreenWindowSceneDelegate methodSignatureForSelector:]):
(-[WKFullscreenWindowSceneDelegate respondsToSelector:]):
(-[WKFullscreenWindowSceneDelegate forwardInvocation:]):
(-[WKFullscreenWindowSceneDelegate forwardingTargetForSelector:]):
(-[WKFullScreenWindowController enterFullScreen:]):
(-[WKFullScreenWindowController beganEnterFullScreenWithInitialFrame:finalFrame:]):
(-[WKFullScreenWindowController _completedExitFullScreen]):
(-[WKFullScreenWindowController _exitFullscreenImmediately]):
(-[WKFullScreenWindowController _dismissFullscreenViewController]):
(-[WKFullScreenWindowController _configureSpatialFullScreenTransition]):
(-[WKFullScreenWindowController _performSpatialFullScreenTransition:completionHandler:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/265416@main">https://commits.webkit.org/265416@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4adb8ca09a297dd2a5892a47b61df1fd7f4fa66c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10823 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11042 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11347 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12473 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10376 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13416 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11021 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13281 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10983 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/14613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/9116 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12876 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9186 "Passed tests") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/17030 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10257 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9925 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13179 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10392 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/8478 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9551 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2592 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13823 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10252 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1333,6 +1485,102 @@ - (void)_applicationDidBecomeActive:(NSNotification*)notification
+
+    auto completion = makeBlockPtr([controller = retainPtr(controller), inWindow = retainPtr(inWindow), originalState = retainPtr(originalState), enter, completionHandler = WTFMove(completionHandler)] (BOOL finished) mutable {
+        if (!finished)
+            return;
```
(Just curious — is it intended to skip the `completionHandler` in the case of `!finished`?)

pxlcoder replied:
Not intentional... I should remove this check as the animation isn't intended to be interruptible. Probably better in a follow-up, to avoid mixing with upstreaming.

whsieh replied:
Yep! Definitely good to keep this upstreaming patch as-is.

pxlcoder committed:
Upstream visionOS fullscreen transition logic
https://bugs.webkit.org/show_bug.cgi?id=258404
rdar://111164029

Reviewed by Wenson Hsieh.

* Source/WebKit/Platform/spi/visionos/MRUIKitSPI.h: Added.
* Source/WebKit/Platform/spi/visionos/RealitySimulationServicesSPI.h: Added.
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(WebKit::useSpatialFullScreenTransition):
(WebKit::resizeScene):
(-[WKFixedSizeWindow _shouldResizeWithScene]):
(-[WKFullScreenParentWindowState initWithWindow:]):
(-[WKFullscreenWindowSceneDelegate initWithController:originalDelegate:]):
(-[WKFullscreenWindowSceneDelegate originalDelegate]):
(-[WKFullscreenWindowSceneDelegate windowScene:shouldCloseForReason:]):
(-[WKFullscreenWindowSceneDelegate methodSignatureForSelector:]):
(-[WKFullscreenWindowSceneDelegate respondsToSelector:]):
(-[WKFullscreenWindowSceneDelegate forwardInvocation:]):
(-[WKFullscreenWindowSceneDelegate forwardingTargetForSelector:]):
(-[WKFullScreenWindowController enterFullScreen:]):
(-[WKFullScreenWindowController beganEnterFullScreenWithInitialFrame:finalFrame:]):
(-[WKFullScreenWindowController _completedExitFullScreen]):
(-[WKFullScreenWindowController _exitFullscreenImmediately]):
(-[WKFullScreenWindowController _dismissFullscreenViewController]):
(-[WKFullScreenWindowController _configureSpatialFullScreenTransition]):
(-[WKFullScreenWindowController _performSpatialFullScreenTransition:completionHandler:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/265416@main
